File: java/ContainsDuplicateII.java
/*
 * Filename: ContainsDuplicateII.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 04 Jun 2015 12:22:26 PM CST
 */

public class Solution {
	public boolean containsNearbyDuplicate(int[] nums, int k) {
		if (nums == null || nums.length <= 1) {
			return false;
		}

		HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
		for (int i = 0; i < nums.length; i++) {
			if (map.containsKey(nums[i])) {
				if (Math.abs(i - map.get(nums[i])) <= k) {
					return true;
				}
			}
			map.put(nums[i], i);
		}

		return false;
	}
}

File: cpp/contains_duplicate_ii.cc
/*
 * Filename: contains_duplicate_ii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 14 Aug 2015 12:46:34 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		bool containsNearbyDuplicate(vector<int>& nums, int k) {
			if (nums.size() <= 1)
			{
				return false;
			}

			map<int, int> mapping;

			for (auto i = 0; i < nums.size(); i++)
			{
				if (mapping.find(nums[i]) != mapping.end() && i - mapping[nums[i]] <= k)
				{
					return true;
				}
				else
				{
					mapping[nums[i]] = i;
				}
			}

			return false;
		}
};
